### PR TITLE
[B+C] Adds GameRule enum. Adds BUKKIT-4316

### PR DESCRIPTION
**Pull request by kezz**
**Justification for this PR**
This PR provides developers with a much needed way of securing the future of game rules in Minecraft. It also provides developers who are new to game rules an easy way to understand and use them in their plugins without having to trawl around the Minecraft wiki for explanations.

**The Breakdown**
This PR adds a new enum, GameRule. This enum maps a game rule to a string and also provides a method for converting strings into GameRules. Each GameRule is fully annotated, meaning developers can easily tell what each enum does.

This PR also adds new methods in (Craft)World that allow developers to use the new enum in place of the string alternative.

**Links**
Bukkit PR - https://github.com/Bukkit/Bukkit/pull/886
CraftBukkit PR - https://github.com/Bukkit/CraftBukkit/pull/1181
Jira Ticket - [BUKKIT-4316](https://bukkit.atlassian.net/browse/BUKKIT-4316)

### CODE REVIEW

**Comment by spaceemotion** [comment]
I like that addition, but I would keep the string methods as they are (no deprecation). Plugin developers can set their own game-rules this way.

Example: An arena plugin can set up a world with a custom gamerule that defines the gamemode. Not very practical, but just an example :P

**Comment by kezz** [comment]
Hmm I see where you are coming from @spaceemotion . I'll un-deprecate now.

**Comment by turt2live** [comment]
You can't do that anyways...

From CraftBukkit's setGameRuleValue(String rule, String value)

```
if (!isGameRule(rule)) return false;
```

Edit: At least that's how it should be... your CB PR removes that line as an undocumented change.

**Comment by kezz** [comment]
Oh yeah. Surprising I didn't realise that considering I've spent the last 30 minutes staring at game mode functions :smile: 

Edit: @turt2live my does not remove that line from `setGameRuleValue(String rule, String value)`. It only does not implement it in `setGameRuleValue(GameRule rule, Stirng value)`

File: src/main/java/org/bukkit/GameRule.java [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+     * @return The GameRule with the given value or {@link null} if it does not exist
+     */
+    public static GameRule getByRule(String rule) {
+        Validate.notNull(rule, "Rule cannot be null");
+        return BY_RULE.get(rule);
+    }
+
+    static {
+        for(GameRule rule : values()) 
+            BY_RULE.put(rule.toString(), rule);
```
Comment (spaceemotion) [inline src/main/java/org/bukkit/GameRule.java]: Getting "Loops must use braces" try adding { and } around ..

Comment (kezz) [inline src/main/java/org/bukkit/GameRule.java]: Fixed :smiley:

File: src/main/java/org/bukkit/GameRule.java [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+        return this.rule;
+    }
+
+    /**
+     * Gets the {@link GameRule} represented by the specified string
+     *
+     * @param rule Value to check
+     * @return The GameRule with the given value or {@link null} if it does not exist
+     */
+    public static GameRule getByRule(String rule) {
```
Comment (kethomassen) [inline src/main/java/org/bukkit/GameRule.java]: `getGameRuleByString` is probably better.

Comment (kezz) [inline src/main/java/org/bukkit/GameRule.java]: I'm going with `getByString` in order to be consistent with things like `ChatColor`.

File: src/main/java/org/bukkit/GameRule.java [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+     */
+    DO_FIRE_TICK("doFireTick"),
+
+    /**
+     * Used to toggle whether mobs drop loot
+     */
+    DO_MOB_LOOT("doMobLoot"),
+
+    /**
+     * Used to toggle the natural spawning of mobs (does not effect mob eggs or mob spawners)
```
Comment (turt2live) [inline src/main/java/org/bukkit/GameRule.java]: We use an 80 column limit on JavaDocs

File: src/main/java/org/bukkit/GameRule.java [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+     */
+    DO_MOB_SPAWNING("doMobSpawning"),
+
+    /**
+     * Used to toggle the dropping of blocks when broken (includes TNT destroying blocks)
```
Comment (turt2live) [inline src/main/java/org/bukkit/GameRule.java]: We use an 80 column limit on JavaDocs

File: src/main/java/org/bukkit/GameRule.java [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+     */
+    DO_TILE_DROPS("doTileDrops"),
+
+    /**
+     * Used to toggle whether players will keep their inventory on death
+     */
+    KEEP_INVENTORY("keepInventory"),
+
+    /**
+     * Used to toggle mobs griefing blocks (e.g creepers blowing up blocks, endermen picking up blocks, etc)
```
Comment (turt2live) [inline src/main/java/org/bukkit/GameRule.java]: We use an 80 column limit on JavaDocs

File: src/main/java/org/bukkit/GameRule.java [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+     */
+    MOB_GRIEFING("mobGriefing"),
+
+    /**
+     * Used to toggle players naturally regaining health regardless of their hunger level
```
Comment (turt2live) [inline src/main/java/org/bukkit/GameRule.java]: We use an 80 column limit on JavaDocs

File: src/main/java/org/bukkit/GameRule.java [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+        Validate.notNull(rule, "Rule cannot be null");
+        return BY_RULE.get(rule);
+    }
+
+    static {
+        for(GameRule rule : values()) { 
+            BY_RULE.put(rule.toString(), rule);
+        }
+    }
+
```
Comment (turt2live) [inline src/main/java/org/bukkit/GameRule.java]: This blank line should be removed

File: src/main/java/org/bukkit/GameRule.java [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+import org.apache.commons.lang.Validate;
+
+import com.google.common.collect.Maps;
+
+/**
+ * Represents the many game rules that can be set with {@link World#setGameRuleValue(GameRule, String)}
+ */
+public enum GameRule {
+    /**
+     * Used to toggle text output of command block commands to the console
```
Comment (turt2live) [inline src/main/java/org/bukkit/GameRule.java]: All of these JavaDocs would be better worded as something like

```
Represents the game rule to _______
```

In this specific example I would use:

```
Represents the game rule to toggle command block output
```

File: src/main/java/org/bukkit/GameRule.java [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+     */
+    COMMAND_BLOCK_OUTPUT("commandBlockOutput"),
```
Comment (turt2live) [inline src/main/java/org/bukkit/GameRule.java:18]: ~~To follow suite with the other enum values, I would recommend `DO_COMMAND_BLOCK_OUTPUT`~~

EDIT: I would actually likely remove the "DO_" from all of these values and re-word them accordingly

File: src/main/java/org/bukkit/GameRule.java [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+import com.google.common.collect.Maps;
+import org.apache.commons.lang.Validate;
+
+import java.util.Map;
+
+/**
+ * Represents the many game rules that can be set with {@link World#setGameRuleValue(GameRule, String)}
+ */
+public enum GameRule {
+    /**
```
Comment (turt2live) [inline src/main/java/org/bukkit/GameRule.java:14]: A blank line should be between `{` and `/**`

File: src/main/java/org/bukkit/World.java [hunk 11]
```diff
@@ -1064,6 +1070,33 @@
...
+     * Get existing rules
+     *
+     * @return An array of {@link GameRule}s
+     */
+    public GameRule[] getExistingGameRules();
+
+    /**
+     * Gets the current state of the specified rule
+     *
+     * @param rule {@link GameRule} to look up value of
```
Comment (turt2live) [inline src/main/java/org/bukkit/World.java]: `The {@link GameRule} to determine the value of`

File: src/main/java/org/bukkit/World.java [hunk 12]
```diff
@@ -1064,6 +1070,33 @@
     public boolean isGameRule(String rule);
 
     /**
+     * Get existing rules
+     *
+     * @return An array of {@link GameRule}s
```
Comment (turt2live) [inline src/main/java/org/bukkit/World.java]: `An array of existing {@link GameRule}s`

File: src/main/java/org/bukkit/World.java [hunk 13]
```diff
@@ -1064,6 +1070,33 @@
     public boolean isGameRule(String rule);
 
     /**
+     * Get existing rules
```
Comment (turt2live) [inline src/main/java/org/bukkit/World.java]: Gets existing rules for this world

File: src/main/java/org/bukkit/World.java [hunk 14]
```diff
@@ -1064,6 +1070,33 @@
...
+    /**
+     * Gets the current state of the specified rule
+     *
+     * @param rule {@link GameRule} to look up value of
+     * @return String value of rule
+     */
+    public String getGameRuleValue(GameRule rule);
+
+    /**
+     * Set the specified gamerule to specified value.
```
Comment (turt2live) [inline src/main/java/org/bukkit/World.java]: Sets a specified GameRule to a specified value

File: src/main/java/org/bukkit/World.java [hunk 15]
```diff
@@ -1064,6 +1070,33 @@
...
+     * <p>
+     * The rule may attempt to validate the value passed, will return true if
```
Comment (turt2live) [inline src/main/java/org/bukkit/World.java]: This may validate the value to ensure it is applicable to the GameRule. If the value was not valid this will return false.

File: src/main/java/org/bukkit/World.java [hunk 16]
```diff
@@ -1064,6 +1070,33 @@
...
+     * value was set.
+     *
+     * @param rule {@link GameRule} to set
```
Comment (turt2live) [inline src/main/java/org/bukkit/World.java]: The GameRule to set

File: src/main/java/org/bukkit/World.java [hunk 17]
```diff
@@ -1064,6 +1070,33 @@
...
+     * @param value Value to set rule to
```
Comment (turt2live) [inline src/main/java/org/bukkit/World.java]: The value to set the GameRule to

File: src/main/java/org/bukkit/World.java [hunk 18]
```diff
@@ -1064,6 +1070,33 @@
...
+     * @return {@code true} if rule was set
```
Comment (turt2live) [inline src/main/java/org/bukkit/World.java]: `@return true if the GameRule was set to the specified value; false otherwise`

File: src/main/java/org/bukkit/GameRule.java [hunk 19]
```diff
@@ -0,0 +1,93 @@
...
+
+    @Override
+    public String toString() {
+        return this.rule;
+    }
+
+    /**
+     * Gets the {@link GameRule} represented by the specified string
+     *
+     * @param rule Value to check
```
Comment (turt2live) [inline src/main/java/org/bukkit/GameRule.java]: `@param rule The rule to look for`
`@return The GameRule for the specified name or null if not found`

File: src/main/java/org/bukkit/GameRule.java [hunk 20]
```diff
@@ -0,0 +1,93 @@
...
+
+    /**
+     * Represents the game rule to toggle whether players will keep their
+     * inventory on death
+     */
+    KEEP_INVENTORY("keepInventory"),
+
+    /**
+     * Represents the game rule to toggle mobs griefing blocks.
+     * For example: creepers blowing up blocks or endermen picking up blocks
```
Comment (turt2live) [inline src/main/java/org/bukkit/GameRule.java]:
```
Represents...
<p>
Example: Creepers exploding or Endermen picking up blocks
```

File: src/main/java/org/bukkit/GameRule.java [hunk 21]
```diff
@@ -0,0 +1,93 @@
...
+
+    /**
+     * Represents the game rule to toggle the natural spawning of mobs.
+     * This does not effect mob eggs or mob spawners
+     */
+    MOB_SPAWNING("doMobSpawning"),
+
+    /**
+     * Represents the game rule to toggle the dropping of blocks when broken.
+     * This includes TNT destroying blocks
```
Comment (turt2live) [inline src/main/java/org/bukkit/GameRule.java:45]:
```
Represents...
<p>
This includes...
```

File: src/main/java/org/bukkit/GameRule.java [hunk 22]
```diff
@@ -0,0 +1,93 @@
+package org.bukkit;
+
+import org.apache.commons.lang.Validate;
+
+import java.util.Map;
+
+import com.google.common.collect.Maps;
+
+/**
+ * Represents the many game rules that can be set with {@link World#setGameRuleValue(GameRule, String)}
```
Comment (turt2live) [inline src/main/java/org/bukkit/GameRule.java]: `An enum to represent the possible game rule values`

File: src/main/java/org/bukkit/World.java [hunk 23]
```diff
@@ -1,17 +1,13 @@
 package org.bukkit;
 
 import java.io.File;
-import org.bukkit.generator.ChunkGenerator;
-import java.util.Collection;
-import java.util.HashMap;
-import java.util.List;
-import java.util.Map;
-import java.util.UUID;
+import java.util.*;
```
Comment (turt2live) [inline src/main/java/org/bukkit/World.java]: This is a no-no.

File: src/main/java/org/bukkit/World.java [hunk 24]
```diff
@@ -1064,6 +1066,34 @@
...
+     * @param rule The {@link GameRule} to determine the value of
+     * @return String value of rule
+     */
+    public String getGameRuleValue(GameRule rule);
+
+    /**
+     * Set a specified {@link GameRule} to specified value.
+     * <p>
+     * This may validate the value to ensure it is applicable to the GameRule.
+     * If the value was not valid this will return {@code false}.
```
Comment (turt2live) [inline src/main/java/org/bukkit/World.java]: Remove the `{@code false}` and use plain text

File: src/main/java/org/bukkit/World.java [hunk 25]
```diff
@@ -1064,6 +1066,34 @@
...
+     *
+     * @param rule The GameRule to set
+     * @param value The value to set the GameRule to
+     * @return {@code true} if the GameRule was set to the specified value;
```
Comment (turt2live) [inline src/main/java/org/bukkit/World.java]: Remove the `{@code true}` and use plain text

**Comment by turt2live** [comment]
Sorry for all the comments... I can only notice so many things at a time :(

**Comment by kezz** [comment]
It's fine :smiley: I'm more than happy to update these

File: src/main/java/org/bukkit/World.java (hunk 23 above)
Comment (kezz) [inline src/main/java/org/bukkit/World.java]: Silly me forgetting to setup IDEA on my laptop XD

File: src/main/java/org/bukkit/World.java [hunk 26]
```diff
@@ -12,6 +11,7 @@
 import org.bukkit.block.Block;
 import org.bukkit.entity.*;
 import org.bukkit.generator.BlockPopulator;
+import org.bukkit.generator.ChunkGenerator;
```
Comment (turt2live) [inline src/main/java/org/bukkit/World.java]: This import moved... 

I'll double check with a core member because this import moved to the correct spot, but it may need to be moved up to where it was. Don't move it just yet though... I'll get back to you on this

Comment (turt2live) [inline src/main/java/org/bukkit/World.java]: Just leave it where it is... ignore me on this one

Comment (kezz) [inline src/main/java/org/bukkit/World.java]: @turt2live done :smiley_cat:

**Comment by turt2live** [comment]
Supplied testing materials: 

[Source + JAR](https://drive.google.com/folderview?id=0B6JIYSBkLkAnbVNDZkw4dXZvbDQ&usp=sharing&tid=0B6JIYSBkLkAnTTZYd3Z4ZTMzV3c) - Instructions included.

**Comment by kezz** [comment]
@turt2live sounds like a JUnit test. Shall I?

**Comment by turt2live** [comment]
If you can make a test easily, go for it. Anything that can be tested with JUnit should. 

Also: Don't forget to address the CraftBukkit side, I left some comments on the diff there.

**Comment by turt2live** [comment]
After some discussion, this PR is being closed for the following reasons, sorry! 
- Game rules are not intended to be mapped to specific values and can change constantly. The current method (string-based) should be the better way to future proof CraftBukkit against added/removed/changed game rules
- Inactivity and inability to be pulled

Thank you for your contribution, but this PR sadly does not meet the requirements for this particular scenario.

**Comment by JeromSar** [comment]
@turt2live If Gamerules are likely to change in the future, than that is all the more reason to have these mapped to an Enum. Similar systems are in place for Material (block IDs are magic values and likely to change) and Color (Dye color bytes are magic values). Following this reasoning, it makes sense that Gamrules should too be mapped to an enum. This allows plugin developers to create more robust plugins by only referencing the enum, which is a more solid implementation of Gamerule values.

Long story short, if Gamerule values are likely to change, then they _too_ are magic values and thus a better API for Gamrules is necessary. The enum implementation @Kezz101 provided is an excellent solution and should not be ignored like it is now.

This pull request should be reopened and reconsidered.
